### PR TITLE
Slashing the production image size by 25%

### PR DESCRIPTION
**Pull request by ThatDeparted2061**
### Reduce the size of production docker image #3142 
# Results
IMG_1746
- The first image i.e. with the before tag is the production image before the improvements ( 700 MB )
- The second image is the production image after the improvements ( 527 MB )
- The build time of both the images is the same ( ~400 seconds )
## Base Image Changes:

- Switched to Alpine-based images for all stages
- Reduced base image size significantly (Ruby slim → Alpine)
- Used Alpine-specific package manager (apk instead of apt-get)

## Build Optimizations:

- Combined ENV statements to reduce layers
- Combined RUN commands to reduce layers
- Used multi-stage builds more effectively
- Removed unnecessary packages after installation
- Used npm ci instead of npm install for more reliable builds

## Dependency Management:

- Separated build and runtime dependencies
- Installed only necessary runtime packages in final stage
- Used --no-cache flag with apk to keep image size down
- Kept essential Ruby gems and removed cache

## Security Improvements:

- Removed development and test dependencies
- Minimized attack surface with Alpine base
- Reduced number of installed packages

## Performance Enhancements:

- Maintained jemalloc configuration for better memory management
- Kept bootsnap precompilation for faster boot times
- Optimized layer caching for faster builds

## Summary by CodeRabbit

- **New Features**
	- Enhanced Docker build process for improved efficiency and performance.
	- Streamlined installation of Ruby and JavaScript dependencies.
  
- **Bug Fixes**
	- Adjusted environment variables for better configuration clarity.

- **Chores**
	- Updated base images to lighter Alpine versions for reduced image size.

### CODE REVIEW

File: Dockerfile.production [hunk 1]
```diff
@@ -1,90 +1,109 @@
...
+# Set production environment variables
+ENV RAILS_ENV=production \
+    NODE_ENV=production \
+    BUNDLE_PATH=/usr/local/bundle \
+    RAILS_LOG_TO_STDOUT=true \
+    BUNDLE_DEPLOYMENT=true \
+    BUNDLE_WITHOUT="development:test" \
+    RAILS_SERVE_STATIC_FILES="true" \
+    LD_PRELOAD="/usr/lib/jemalloc.so.2" \
+    MALLOC_CONF="dirty_decay_ms:1000,narenas:2,background_thread:true"
```
[the remark this answers is not in the document]
Comment (ThatDeparted2061) [reply Dockerfile.production:85]: using jemalloc doesn't shoot me any error either.
but if its a mandatory thing then i'd do it 
@VaibhavUpreti

[the remark this answers is not in the document]
Comment (VaibhavUpreti) [reply Dockerfile.production:85]: Yes we need to configure jemalloc for memory optimisation.

[the remark this answers is not in the document]
Comment (ThatDeparted2061) [reply Dockerfile.production:85]: So shud i change it to libjemalloc ?

[the remark this answers is not in the document]
Comment (ThatDeparted2061) [reply Dockerfile.production:85]: I am sorry but I didnt quite catch what you meant

**Comment by tachyons** [comment]
Thanks for the great work. I've closed the original issue due to potential issues it can cause in future.

**Comment by ThatDeparted2061** [comment]
@tachyons kindly don't mind the me requesting a reply on this matter on slack - Harsh Rao

**Pull request closed without merging**